src/centralized.rs

```diff
@@ -0,0 +1,499 @@
+        }
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(data);
+        utils::oracle_binomial(&digest, setup.q)
+    }
+
+    /// Depth-first search which goes through all potential round candidates
+    /// and returns first round candidate Round{t, x_1, ..., x_u)} such that:
+    /// - for all i ∈ [0, u-1], H0(x_i+1) ∈ bins[H1(t, x_1, ..., x_i)]
+    /// - H2(t, x_0, ..., x_u) = true
+    fn dfs(
+        setup: &Setup,
+        bins: &Vec<Vec<Data>>,
+        round: &Round,
+        nb_steps: Arc<AtomicUsize>,
+    ) -> Option<Proof> {
+        if round.s_list.len() == setup.u {
+            if Proof::h2(setup, round) {
+                let r = round.v;
+                let d = round.t;
+                let items = round.s_list.clone();
+                return Some(Proof { r, d, items });
+            } else {
+                return None;
+            }
+        }
+        let result = bins[round.h_usize].par_iter().find_map_first(|&s| {
+            if nb_steps.load(atomic::Ordering::Relaxed) == setup.d {
+                return None;
+            }
+            nb_steps.fetch_add(1, atomic::Ordering::Relaxed);
+            Self::dfs(setup, bins, &Round::update(round, s), nb_steps.clone())
+        });
+        result
+    }
+
+    /// Indexed proving algorithm, returns an empty proof if no suitable
+    /// candidate is found within the setup.b steps.
+    fn prove_index(setup: &Setup, set: &[Data], v: usize) -> (usize, Option<Proof>) {
+        let mut bins: Vec<Vec<Data>> = vec![];
+        for _ in 0..setup.n_p {
+            bins.push(vec![]);
+        }
+        for &s in set.iter() {
+            bins[Proof::h0(setup, v, s)].push(s);
+        }
+        let nb_steps = Arc::new(AtomicUsize::new(0));
+        for t in 0..setup.d {
+            if nb_steps.load(atomic::Ordering::Relaxed) == setup.b {
+                return (0, None);
+            }
+            nb_steps.fetch_add(1, atomic::Ordering::Relaxed);
+            let round = Round::new(v, t + 1, setup.n_p);
+            let res = Proof::dfs(setup, &bins, &round, nb_steps.clone());
+            if res.is_some() {
+                return (nb_steps.load(atomic::Ordering::Relaxed), res);
+            }
+        }
+        (nb_steps.load(atomic::Ordering::Relaxed), None)
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
+    /// Calls up to setup.r times the prove_index function and returns an empty
+    /// proof if no suitable candidate is found.
+    pub fn prove(setup: &Setup, set: &[Data]) -> Self {
+        for v in 0..setup.r {
+            if let (_, Some(proof)) = Proof::prove_index(setup, set, v + 1) {
+                return proof;
+            }
+        }
+        Proof::new()
+    }
+
+    /// Alba's proving algorithm used for benchmarking, returning a proof as
+    /// well as the number of  steps ran to find it.
+    pub fn bench(setup: &Setup, set: &[Data]) -> (usize, usize, Self) {
```

Comment:
Can you explain me more how to have the internal prove benched if it is not public?

As for calling the prove function on the benched, I have no problem with it, I just wanted to avoid the extra accumulator overhead.